packages-nextgen/kestrel_core/src/kestrel/cache/sqlite.py

```diff
@@ -0,0 +1,151 @@
+from kestrel.ir.instructions import (
+    Construct,
+    Filter,
+    Instruction,
+    Return,
+    SourceInstruction,
+    TransformingInstruction,
+    Variable,
+)
+
+_logger = logging.getLogger(__name__)
+
+
+@typechecked
+class SqliteTranslator(SqlTranslator):
+    def __init__(self):
+        super().__init__(
+            sqlite.dialect(), dt_parser, "time"
+        )  # FIXME: need mapping for timestamp?
+
+
+@typechecked
+def _set_from(translator: SqliteTranslator, table_name: str):
+    """Utility method to set the table for a SQL query.
+
+    This is used to handle Construct instructions.
+    """
+    translator.query = translator.query.select_from(table(table_name))
+
+
+@typechecked
+class SqliteCache(Cache):
+    def __init__(
+        self,
+        initial_cache: Optional[Mapping[UUID, DataFrame]] = None,
+        session_id: Optional[UUID] = None,
+    ):
+        super().__init__()
+
+        basename = self.session_id or "cache"
+        path = f"{basename}.db"
+
+        # for an absolute file path, the three slashes are followed by the absolute path
+        # for a relative path, it's also three slashes?
+        self.engine = create_engine(f"sqlite:///{path}")
+        self.connection = self.engine.connect()
+
+        if initial_cache:
+            for instruction_id, data in initial_cache.items():
+                self.store(instruction_id, data)
+
+    def __del__(self):
+        self.connection.close()
+
+    def __getitem__(self, instruction_id: UUID) -> DataFrame:
+        result = read_sql(instruction_id.hex, self.connection)
+        return result
+
+    def __delitem__(self, instruction_id: UUID):
+        self.connection.execute(text(f'DROP TABLE "{instruction_id.hex}"'))
+
+    def _to_sql(self, df: DataFrame, name: str):
+        df.to_sql(name, con=self.connection, if_exists="replace", index=False)
+
+    def store(
+        self,
+        instruction_id: UUID,
+        data: DataFrame,
+    ):
+        self._to_sql(data, instruction_id.hex)
+
+    def evaluate_graph(
+        self,
+        graph: IRGraphSoleInterface,
+        instructions_to_evaluate: Optional[Iterable[Instruction]] = None,
+    ) -> Mapping[UUID, DataFrame]:
+        if not instructions_to_evaluate:
+            instructions_to_evaluate = graph.get_sink_nodes()
+        stacks = []
+        for inst in instructions_to_evaluate:
+            stacks.append(self._visit(graph, inst))
+        mapping = {}
+        for stack in stacks:
+            tail = stack[-1]
+            mapping[tail.id] = self._evaluate_instruction_stack(stack)
+        _logger.debug("mapping: %s", mapping)
+        return mapping
+
+    def _visit(
+        self,
+        graph: IRGraphSoleInterface,
+        instruction: Instruction,
+        stack: Optional[List[Instruction]] = None,
+    ) -> List[Instruction]:
+        """
+        Depth-first visit of nodes in graph from variable to variable or variable to source
```

Comment:
Do we assume the graph in input is a linear graph, so `next(graph.predecessors(instruction))` will cover all nodes backward?

1. If we have something like the following (`c` is a Filter node with reference to multiple variables):
```
a ->b; b->c; c->d; e->f; f->c

# this is coming from
a = NEW ...
e = NEW ...
d = a WHERE xxx = e.pid
```
We need to think about ways to deal with it.

2. If it is just a linear graph, why not call `translator.add_instruction(instruction)` from the sink node backward until any source node (`SourceInstruction` or `Variable`). This may result in combining `_visit()` and `_evaluate_instruction_stack` into one recursive function like:
```
def _evaluate_linear_instruction_graph(self, graph, instruction, translator):
    if `SourceInstruction` or `Variable` or `Ret`:
        ... # same as current _evaluate_instruction_stack()
    elif isinstance(instruction, TransformingInstruction):
        translator.add_instruction(instruction)
        _evaluate_linear_instruction_graph(self, graph, next(graph.predecessors(instruction)), translator)
```